config.org
#+TITLE: Light-Emacs
#+AUTHOR: Hitendra (hi720tu)
#+Description: My personal configuration of GNU Emacs.

* Table of Contents:toc_3:
- [[#package-manager][Package Manager]]
  - [[#installing-straightel][Installing straight.el]]
- [[#look--feel][Look & Feel]]
  - [[#fonts][Fonts]]
  - [[#frame-boarders-and-window-dividers][Frame boarders and Window dividers]]
  - [[#org][org]]
- [[#packages][Packages]]
  - [[#treesit-auto][treesit-auto]]
  - [[#markdown-mode][Markdown-mode]]
  - [[#built-in-packages][Built-in Packages]]
  - [[#org-packages][Org Packages]]
    - [[#org-roam][Org-Roam]]
    - [[#org-roam-ui][org-roam-ui]]
    - [[#toc-org][TOC-org]]
    - [[#org-modern][org-modern]]
    - [[#org-modern-indent][org-modern-indent]]
  - [[#magit][Magit]]
  - [[#minibuffer-related][Minibuffer related]]
    - [[#vertico][Vertico]]
    - [[#orderless][Orderless]]
    - [[#marginalia][Marginalia]]
  - [[#emmet][Emmet]]
  - [[#indentation-line][Indentation line]]
- [[#improved-default-settings][Improved Default Settings]]
  - [[#auto-save--backup-files-management][Auto-save & Backup files management]]
  - [[#changing-default-encoding-to-utf-8][Changing default encoding to UTF-8]]

* Package Manager

** Installing straight.el

#+begin_src emacs-lisp

  ;; (defvar bootstrap-version)
  ;; (let ((bootstrap-file
  ;;        (expand-file-name
  ;;         "straight/repos/straight.el/bootstrap.el"
  ;;         (or (bound-and-true-p straight-base-dir)
  ;;             user-emacs-directory)))
  ;;       (bootstrap-version 7))
  ;;   (unless (file-exists-p bootstrap-file)
  ;;     (with-current-buffer
  ;;         (url-retrieve-synchronously
  ;;          "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
  ;;          'silent 'inhibit-cookies)
  ;;       (goto-char (point-max))
  ;;       (eval-print-last-sexp)))
  ;;   (load bootstrap-file nil 'nomessage))

  ;; (straight-use-package 'use-package)
  ;; (setq straight-use-package-by-default t)

  ;; ;; Official recommendation
  ;; (setq package-enable-at-startup nil)

#+end_src

* Look & Feel

#+begin_src emacs-lisp

  ;; Clean interface
  (tool-bar-mode -1)
  (scroll-bar-mode -1)

  ;; Adding relative line numbers
  (global-display-line-numbers-mode 1)
  (setq display-line-numbers-type 'relative)

  ;; Vertical window split as default
  (setq split-height-threshold nil)
  (setq split-width-threshold 0)

  ;; Smooth scrolling
  (pixel-scroll-precision-mode t)
  (setq scroll-step 1
        scroll-conservatively 10000
        scroll-margin 5)

  ;; Line highlight
  (global-hl-line-mode 1)

  ;; Matching parentheses color
  ;; (set-face-attribute 'show-paren-match nil :foreground "#02f543" :background "#02f543")

  ;; Line Spacing
  (setq-default line-spacing 0.12)

  ;; Main theme
  (setq modus-themes-headings
        '((1 . (variable-pitch 1.5))
  	(2 . (variable-pitch 1.3))
  	(3 . (variable-pitch 1.1))
  	(t . (variable-pitch 1.0))))
  (load-theme 'modus-operandi-tinted t)

#+end_src

** Fonts

#+begin_src emacs-lisp

  (set-face-attribute 'default nil :family "Iosevka")
  (set-face-attribute 'fixed-pitch nil :family "Iosevka")
  ;; (set-face-attribute 'variable-pitch nil :family "Libre Baskerville")
  (set-face-attribute 'variable-pitch nil :family "Iosevka Aile")

  (set-fontset-font t 'telugu "NTR")

#+end_src

** Frame boarders and Window dividers

Taken from org-modern GitHub repo: [[Ehttps://github.com/minad/org-modern?tab=readme-ov-file#configuration][Link]]
#+begin_src emacs-lisp

  ;; (modify-all-frames-parameters
  ;;  '((right-divider-width . 40)
  ;;    (internal-border-width . 40)))
  ;; (dolist (face '(window-divider
  ;; 		window-divider-first-pixel
  ;; 		window-divider-last-pixel))
  ;;   (face-spec-reset-face face)
  ;;   (set-face-foreground face (face-attribute 'default :background)))
  ;; (set-face-background 'fringe (face-attribute 'default :background))

#+end_src

** org

#+begin_src emacs-lisp

  (add-hook 'org-mode-hook #'visual-line-mode)

  ;; (set-face-attribute 'org-link nil :underline nil)

  (setq
   ;; Settings
   org-auto-align-tags nil
   org-tags-column 0
   org-catch-invisible-edits 'show-and-error
   org-special-ctrl-a/e t
   org-insert-heading-respect-content t
   org-startup-indented t
   org-startup-truncated nil
   ;; Kanban workflow
   org-todo-keywords '((sequence "IDEA" "RESEARCH" "TODO" "WORK-IN-PROGRESS" "|" "DONE" "DROPPED"))
   
   ;; Styling
   org-indent-indentation-per-level 2
   org-hide-emphasis-markers t
   org-pretty-entities t
   org-agenda-tags-column 0
   org-ellpisis "…")

  ;; Extra styling for code block
  (let ((med-col "#a60000")
        (re-tint "#fbe9ef"))
    (custom-set-faces
     `(org-block-begin-line ((t :background ,re-tint :foreground ,med-col :extend t)))
     `(org-block-end-line ((t :background ,re-tint :foreground ,med-col :extend t)))
     `(org-modern-block-name ((t :background ,med-col :foreground "white" :weight bold)))
     `(org-modern-block-keyword ((t :background ,re-tint :foreground ,med-col :weight bold)))))

#+end_src


* Packages

** treesit-auto

#+begin_src emacs-lisp

  ;; (use-package treesit-auto
  ;;   :ensure t
  ;;   :custom (treesit-auto-install 'prompt)
  ;;   :config (treesit-auto-add-to-auto-mode-alist 'all)
  ;;   (global-treesit-auto-mode))

#+end_src

** Markdown-mode

#+begin_src emacs-lisp

  (use-package markdown-mode
    :ensure t
    :mode ("README\\.md\\'" . gfm-mode)
    :init (setq markdown-command "multimarkdown")
    :bind (:map markdown-mode-map
  	      ("C-c C-e" . markdown-do)))

#+end_src

** Built-in Packages

#+begin_src emacs-lisp

  ;; Enabling rainbow-mode
  (use-package rainbow-mode
    :ensure t
    :hook (css-mode web-mode prog-mode))

#+end_src

** Org Packages

*** Org-Roam

#+begin_src emacs-lisp

  (use-package org-roam
  :ensure t
  :custom
  (org-roam-directory "~/Documents/RoamNotes")
  (org-roam-node-display-template
   (concat "${title:*} " (propertize "${tags:10}" 'face 'font-lock-keyword-face)))
  :bind (("C-c n t" . org-roam-buffer-toggle)
         ("C-c n f" . org-roam-node-find)
         ("C-c n i" . org-roam-node-insert)
         ("C-c n c" . org-roam-node-capture)
         ("C-c n g" . org-roam-graph))
  :config
  (org-roam-setup))

  (setq org-return-follows-link t)

#+end_src

*** org-roam-ui

#+begin_src emacs-lisp

  (use-package org-roam-ui
  :straight
    (:host github :repo "org-roam/org-roam-ui" :branch "main" :files ("*.el" "out"))
    :after org-roam
    :hook (after-init . org-roam-ui-mode)
    :config
    (setq org-roam-ui-sync-theme t
          org-roam-ui-follow t
          org-roam-ui-update-on-save t
          org-roam-ui-open-on-start t))

#+end_src

*** TOC-org

#+begin_src emacs-lisp

  (use-package toc-org
    :ensure t
    :hook (org-mode . toc-org-mode))

#+end_src

*** org-modern

#+begin_src emacs-lisp

  (use-package org-modern
    :ensure t
    :config (global-org-modern-mode 1)
    (setq org-modern-star 'replace)
    (setq org-modern-hide-stars nil))

#+end_src

*** org-modern-indent

#+begin_src emacs-lisp

  (use-package org-modern-indent
    :straight (org-modern-indent :type git :host github :repo "jdtsmith/org-modern-indent")
    :init (setq org-startup-indented t)
    :config (add-hook 'org-mode-hook #'org-modern-indent-mode 90))

#+end_src

** Magit

#+begin_src emacs-lisp

  (use-package magit
    :ensure t
    :bind (("C-x g" . magit-status)))

#+end_src

** Minibuffer related

*** Vertico

#+begin_src emacs-lisp

  (use-package vertico
    :ensure t
    :init (vertico-mode))

#+end_src

*** Orderless

#+begin_src emacs-lisp

  (use-package orderless
    :ensure t
    :custom
    (completion-styles '(orderless basic))
    (completion-category-overrides '((file (styles partial-completion))))
    (completion-pcm-leading-wildcard t)) ;; Emacs 31: partial-completion behaves like substring

#+end_src

*** Marginalia

#+begin_src emacs-lisp

  (use-package marginalia
    :ensure t
    :init (marginalia-mode))

#+end_src

** Emmet

#+begin_src emacs-lisp

  (use-package emmet-mode
    :ensure t
    :hook ((html-mode . emmet-mode)
  	 (css-mode . emmet-mode)
  	 (web-mode . emmet-mode)))

#+end_src

** Indentation line

#+begin_src emacs-lisp

  (use-package highlight-indent-guides
    :straight t
    :commands highlight-indent-guides-mode
    :hook (prog-mode-hook . highlight-indent-guides-mode)
    :config (setq highlight-indent-guides-method 'character
  		highlight-indent-guides-responsive t))

#+end_src

* Improved Default Settings

** Auto-save & Backup files management

#+begin_src emacs-lisp

  (defvar my-backup-dir "~/.emacs.d/backups/")
  (defvar my-auto-save-dir "~/.emacs.d/auto-saves/")

  (unless (file-exists-p my-backup-dir)
    (make-directory my-backup-dir t))
  (unless (file-exists-p my-auto-save-dir)
    (make-directory my-auto-save-dir t))

  (setq backup-directory-alist `(("." . ,my-backup-dir))
        auto-save-file-name-transforms `((".*" ,my-auto-save-dir t)))

  ;; Setup
  (setq delete-old-versions t
        kept-new-versions 6
        kept-old-versions 2
        version-control t)

#+end_src

** Changing default encoding to UTF-8

#+begin_src emacs-lisp

  (set-language-environment "UTF-8")

#+end_src



config.el
;; (defvar bootstrap-version)
;; (let ((bootstrap-file
;;        (expand-file-name
;;         "straight/repos/straight.el/bootstrap.el"
;;         (or (bound-and-true-p straight-base-dir)
;;             user-emacs-directory)))
;;       (bootstrap-version 7))
;;   (unless (file-exists-p bootstrap-file)
;;     (with-current-buffer
;;         (url-retrieve-synchronously
;;          "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
;;          'silent 'inhibit-cookies)
;;       (goto-char (point-max))
;;       (eval-print-last-sexp)))
;;   (load bootstrap-file nil 'nomessage))

;; (straight-use-package 'use-package)
;; (setq straight-use-package-by-default t)

;; ;; Official recommendation
;; (setq package-enable-at-startup nil)

;; Clean interface
(tool-bar-mode -1)
(scroll-bar-mode -1)

;; Adding relative line numbers
(global-display-line-numbers-mode 1)
(setq display-line-numbers-type 'relative)

;; Vertical window split as default
(setq split-height-threshold nil)
(setq split-width-threshold 0)

;; Smooth scrolling
(pixel-scroll-precision-mode t)
(setq scroll-step 1
      scroll-conservatively 10000
      scroll-margin 5)

;; Line highlight
(global-hl-line-mode 1)

;; Matching parentheses color
;; (set-face-attribute 'show-paren-match nil :foreground "#02f543" :background "#02f543")

;; Line Spacing
(setq-default line-spacing 0.12)

;; Main theme
(setq modus-themes-headings
      '((1 . (variable-pitch 1.5))
	(2 . (variable-pitch 1.3))
	(3 . (variable-pitch 1.1))
	(t . (variable-pitch 1.0))))
(load-theme 'modus-operandi-tinted t)

(set-face-attribute 'default nil :family "Iosevka")
(set-face-attribute 'fixed-pitch nil :family "Iosevka")
;; (set-face-attribute 'variable-pitch nil :family "Libre Baskerville")
(set-face-attribute 'variable-pitch nil :family "Iosevka Aile")

(set-fontset-font t 'telugu "NTR")

;; (modify-all-frames-parameters
;;  '((right-divider-width . 40)
;;    (internal-border-width . 40)))
;; (dolist (face '(window-divider
;; 		window-divider-first-pixel
;; 		window-divider-last-pixel))
;;   (face-spec-reset-face face)
;;   (set-face-foreground face (face-attribute 'default :background)))
;; (set-face-background 'fringe (face-attribute 'default :background))

(add-hook 'org-mode-hook #'visual-line-mode)

;; (set-face-attribute 'org-link nil :underline nil)

(setq
 ;; Settings
 org-auto-align-tags nil
 org-tags-column 0
 org-catch-invisible-edits 'show-and-error
 org-special-ctrl-a/e t
 org-insert-heading-respect-content t
 org-startup-indented t
 org-startup-truncated nil
 ;; Kanban workflow
 org-todo-keywords '((sequence "IDEA" "RESEARCH" "TODO" "WORK-IN-PROGRESS" "|" "DONE" "DROPPED"))
 
 ;; Styling
 org-indent-indentation-per-level 2
 org-hide-emphasis-markers t
 org-pretty-entities t
 org-agenda-tags-column 0
 org-ellpisis "…")

;; Extra styling for code block
(let ((med-col "#a60000")
      (re-tint "#fbe9ef"))
  (custom-set-faces
   `(org-block-begin-line ((t :background ,re-tint :foreground ,med-col :extend t)))
   `(org-block-end-line ((t :background ,re-tint :foreground ,med-col :extend t)))
   `(org-modern-block-name ((t :background ,med-col :foreground "white" :weight bold)))
   `(org-modern-block-keyword ((t :background ,re-tint :foreground ,med-col :weight bold)))))

;; (use-package treesit-auto
;;   :ensure t
;;   :custom (treesit-auto-install 'prompt)
;;   :config (treesit-auto-add-to-auto-mode-alist 'all)
;;   (global-treesit-auto-mode))

(use-package markdown-mode
  :ensure t
  :mode ("README\\.md\\'" . gfm-mode)
  :init (setq markdown-command "multimarkdown")
  :bind (:map markdown-mode-map
	      ("C-c C-e" . markdown-do)))

;; Enabling rainbow-mode
(use-package rainbow-mode
  :ensure t
  :hook (css-mode web-mode prog-mode))

(use-package org-roam
:ensure t
:custom
(org-roam-directory "~/Documents/RoamNotes")
(org-roam-node-display-template
 (concat "${title:*} " (propertize "${tags:10}" 'face 'font-lock-keyword-face)))
:bind (("C-c n t" . org-roam-buffer-toggle)
       ("C-c n f" . org-roam-node-find)
       ("C-c n i" . org-roam-node-insert)
       ("C-c n c" . org-roam-node-capture)
       ("C-c n g" . org-roam-graph))
:config
(org-roam-setup))

(setq org-return-follows-link t)

(use-package org-roam-ui
:straight
  (:host github :repo "org-roam/org-roam-ui" :branch "main" :files ("*.el" "out"))
  :after org-roam
  :hook (after-init . org-roam-ui-mode)
  :config
  (setq org-roam-ui-sync-theme t
        org-roam-ui-follow t
        org-roam-ui-update-on-save t
        org-roam-ui-open-on-start t))

(use-package toc-org
  :ensure t
  :hook (org-mode . toc-org-mode))

(use-package org-modern
  :ensure t
  :config (global-org-modern-mode 1)
  (setq org-modern-star 'replace)
  (setq org-modern-hide-stars nil))

(use-package org-modern-indent
  :straight (org-modern-indent :type git :host github :repo "jdtsmith/org-modern-indent")
  :init (setq org-startup-indented t)
  :config (add-hook 'org-mode-hook #'org-modern-indent-mode 90))

(use-package magit
  :ensure t
  :bind (("C-x g" . magit-status)))

(use-package vertico
  :ensure t
  :init (vertico-mode))

(use-package orderless
  :ensure t
  :custom
  (completion-styles '(orderless basic))
  (completion-category-overrides '((file (styles partial-completion))))
  (completion-pcm-leading-wildcard t)) ;; Emacs 31: partial-completion behaves like substring

(use-package marginalia
  :ensure t
  :init (marginalia-mode))

(use-package emmet-mode
  :ensure t
  :hook ((html-mode . emmet-mode)
	 (css-mode . emmet-mode)
	 (web-mode . emmet-mode)))

(use-package highlight-indent-guides
  :straight t
  :commands highlight-indent-guides-mode
  :hook (prog-mode-hook . highlight-indent-guides-mode)
  :config (setq highlight-indent-guides-method 'character
		highlight-indent-guides-responsive t))

(defvar my-backup-dir "~/.emacs.d/backups/")
(defvar my-auto-save-dir "~/.emacs.d/auto-saves/")

(unless (file-exists-p my-backup-dir)
  (make-directory my-backup-dir t))
(unless (file-exists-p my-auto-save-dir)
  (make-directory my-auto-save-dir t))

(setq backup-directory-alist `(("." . ,my-backup-dir))
      auto-save-file-name-transforms `((".*" ,my-auto-save-dir t)))

;; Setup
(setq delete-old-versions t
      kept-new-versions 6
      kept-old-versions 2
      version-control t)

(set-language-environment "UTF-8")
